### PR TITLE
chore: set up changesets

### PR DESCRIPTION
**Pull request by emmenko**
https://github.com/atlassian/changesets

TL;DR: changesets are a replacement of lerna changelog, as they provide a more convenient way of writing changelogs incrementally. In fact, the tool is primarily built to work in monorepositories.

Additionally, using the bot and the github action, the workflow of releasing becomes almost fully automated.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -157,6 +127,7 @@ jobs:
           ZEIT_NOW_PLAYGROUND: ${{ secrets.ZEIT_NOW_PLAYGROUND }}
 
       - name: Preparing tarballs of packages for testing templates installation
+        if: github.ref != 'refs/heads/master'
```
Comment (emmenko) [inline .github/workflows/main.yml:130]: Unrelated, but there no need to run the starter installation on master builds

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -170,6 +141,7 @@ jobs:
       # - https://github.com/actions/upload-artifact/issues/5
       # - https://github.com/actions/upload-artifact/issues/45
       - name: Installing and building template starter
+        if: github.ref != 'refs/heads/master'
```
Comment (emmenko) [inline .github/workflows/main.yml:144]: Unrelated, but there no need to run the starter installation on master builds

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -27,34 +23,8 @@ jobs:
 
     steps:
       - name: Checkout
-        if: github.ref != 'refs/heads/master'
         uses: actions/checkout@v2
 
-      - name: Checkout
```
Comment (emmenko) [inline .github/workflows/main.yml:33]: This is moved to the other "Release" workflow

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -230,8 +202,3 @@ jobs:
   #         artifact_id: ${{ steps.get_artifact_id.outputs.artifact_id }}
   #       env:
   #         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-      - name: Publishing to npm registry
```
Comment (emmenko) [inline .github/workflows/main.yml:233]: Moved to the other "Release" workflow

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+      - name: Create Release Pull Request or Publish to npm
+        id: changesets
+        # uses: changesets/action@master
+        uses: emmenko/action@nm-fix-npmrc-overwrite
+        with:
+          publish: yarn release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (emmenko) [inline .github/workflows/release.yml:76]: This action does 2 things:
- if there are changesets files in the `.changesets` folder, it means that it's not the time to do a release. Therefore, a "Version packages" PR is opened by the bot to keep track of the changes and the packages that would be updated.
- if there are no changesets, the bot attempts to release the packages. Packages with a new "unreleased" version are effectively published.

The second step is basically triggered by merging the PR created from the first step.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+      # Publish canary releases only if the packages weren't published already
+      - name: Publishing canary releases to npm registry
+        if: steps.changesets.outputs.published != 'true'
+        run: yarn release:canary
```
Comment (emmenko) [inline .github/workflows/release.yml:81]: If the current build does not publish any package, we attempt to publish them as canary.

File: package.json [hunk 7]
```diff
@@ -22,10 +21,12 @@
...
     "format:graphql": "prettier --write --parser graphql '**/*.graphql'",
     "start": "yarn playground:start",
-    "prerelease": "yarn auth && yarn build && yarn typecheck",
+    "changeset": "changeset",
+    "version-packages": "changeset version",
+    "prerelease": "yarn auth && yarn build",
     "prerelease:alpha": "yarn prerelease",
     "prerelease:canary": "yarn prerelease",
-    "release": "lerna publish --exact --dist-tag next",
+    "release": "lerna exec --scope '@commercetools-{frontend,backend}/*' -- npm publish --dist-tag next",
```
Comment (emmenko) [inline package.json:29]: This is due to a current "limitation" of changesets.

More info here: https://github.com/atlassian/changesets/issues/356

File: package.json [hunk 8]
```diff
@@ -22,10 +21,12 @@
     "format:yaml": "prettier --write --parser yaml '**/*.{yaml, yml}'",
     "format:graphql": "prettier --write --parser graphql '**/*.graphql'",
     "start": "yarn playground:start",
-    "prerelease": "yarn auth && yarn build && yarn typecheck",
```
Comment (emmenko) [inline package.json:25]: Not sure why we run `yarn typecheck` during a release, it should not be necessary.

**Comment by emmenko** [comment]
> Wanted changesets for dompurify and flopflip too

🎉

For uikit this will be very useful, given the "broken" changelogs that we have now (see https://github.com/commercetools/ui-kit/issues/1317).

**Comment by tdeekens** [comment]
This looks good to me actually. Anything missing. A way to test it? Or merge and try a probe release?

**Comment by emmenko** [comment]
I already played around a bit with it today in the docs-kit. I haven't done a release yet. The first thing to do is to merge a PR with a changeset and wait for the gh action to open the "Version packages" PR.

**Comment by emmenko** [comment]
Btw, one thing that will be slightly different is that packages won't have a "locked" version, like Lerna.

I don't see it as a big issue to be honest, as long as the packages work and are well documented.

**Comment by tdeekens** [comment]
By locked your mean sync versions in releases of dependencies among which arent locked?

**Comment by emmenko** [comment]
> one thing that will be slightly different is that packages won't have a "locked" version, like Lerna

Actually there might be a way to do that, by using `linked` packages: https://github.com/atlassian/changesets/blob/master/docs/linked-packages.md
I haven't fully understood yet how that works but that might be what we would need.

**Comment by emmenko** [comment]
I suppose something like this:

```
{
  "$schema": "https://unpkg.com/@changesets/config@1.0.3/schema.json",
  "changelog": [
    "@changesets/changelog-github",
    {
      "repo": "commercetools/merchant-center-application-kit"
    }
  ],
  "commit": false,
  "linked": [
    [
      "@commercetools-backend/express",
      "@commercetools-backend/loggers",
      "@commercetools-frontend/actions-global",
      "@commercetools-frontend/application-components",
      "@commercetools-frontend/application-shell",
      "@commercetools-frontend/application-shell-connectors",
      "@commercetools-frontend/assets",
      "@commercetools-frontend/babel-preset-mc-app",
      "@commercetools-frontend/browser-history",
      "@commercetools-frontend/constants",
      "@commercetools-frontend/create-mc-app",
      "@commercetools-frontend/eslint-config-mc-app",
      "@commercetools-frontend/i18n",
      "@commercetools-frontend/jest-preset-mc-app",
      "@commercetools-frontend/jest-stylelint-runner",
      "@commercetools-frontend/l10n",
      "@commercetools-frontend/mc-dev-authentication",
      "@commercetools-frontend/mc-html-template",
      "@commercetools-frontend/mc-http-server",
      "@commercetools-frontend/mc-scripts",
      "@commercetools-frontend/notifications",
      "@commercetools-frontend/permissions",
      "@commercetools-frontend/react-notifications",
      "@commercetools-frontend/sdk",
      "@commercetools-frontend/sentry",
      "@commercetools-frontend/url-utils"
    ]
  ],
  "access": "restricted",
  "baseBranch": "master"
}
```

**Pull request merged**